### PR TITLE
Ballistics do more environmental damage

### PR DESCRIPTION
**Pull request by Lufferly**
## About The Pull Request

Gives ballistics (specifically bullets with the BULLET armor_type, so no lasers) some additional functionality with environmental objects if they do enough damage (above 8)

- Windows, consoles, and vending machines take extra damage from bullets
- Instantly destroys APCs, if the APCs powernet has some charge it will also shock people in a 2 tile radius (extra if the electricity jumps, but that makes it do less damage)
- Bullets will cut random wires if they hit an airlock
- Bullets will trigger and then outright destroy fire alarms
- Bullets will set air alarms to a random mode, then outright destroy them
- Bullets will knock down intercoms
- Bullets will outright destroy security cameras

Also added tips for some of these.

## Why It's Good For The Game

Ive seen a couple people say that ballistics should be syndicate sided because they do more damage to the station, and while I though it was an interesting idea its not really supported by any game mechanics. Now station sided crew have more of a reason to use laser weaponry, while those who have access to ballistics can cause loads of station damage. I also think the it adds a lot of interesting interactions other than just shooting people with a gun till they die. Traitors could shoot an APC out to shut all local airlocks off or shock those near it. Or they could shoot an airlock a couple of times to lock pursuers off, or shoot a fire alarm on their way past it. It gives them a lot more options to create environmental destruction, which is also interesting for all crewmembers.

## Changelog

:cl: Seven
balance: Physical bullets do extra damage to windows, consoles, and vending machines.
add: Physical bullets now have some special interactions with some common machinery, along with outright destroying some!
add: This includes APCs, fire alarms, air alarms, airlocks, intercoms, and security cameras.
/:cl:

### CODE REVIEW

File: code/game/machinery/camera/camera.dm [hunk 1]
```diff
@@ -578,3 +577,15 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/camera/xray, 0)
...
+/obj/machinery/camera/bullet_act(obj/projectile/bullet)
+	// Energy attacks dont do anything special
+	if (bullet.armor_flag != BULLET)
+		return ..()
+
+	if (bullet.damage >= BULLET_MINIMUM_DAMAGE_NEEDED_TO_EXECUTE_MACHINERY)
+		do_sparks(number = 2, cardinal_only = FALSE, source = src)
+		atom_break()
+
+	return ..()
```
Comment (Fikou) [inline code/game/machinery/camera/camera.dm:591]: early returning here is pretty silly if you do the same thing
```suggestion
	// Energy attacks dont do anything special
	if (bullet.armor_flag == BULLET && bullet.damage >= BULLET_MINIMUM_DAMAGE_NEEDED_TO_EXECUTE_MACHINERY)
		do_sparks(number = 2, cardinal_only = FALSE, source = src)
		atom_break()

	return ..()
```

File: code/game/machinery/doors/airlock.dm [hunk 2]
```diff
@@ -1591,6 +1591,23 @@
...
+	if (bullet.armor_flag != BULLET)
+		return ..()
+
+	// Dont cut the wires if the airlock has been reinforced with something above iron
+	if (bullet.damage > BULLET_MINIMUM_DAMAGE_NEEDED_TO_EXECUTE_MACHINERY && security_level <= AIRLOCK_SECURITY_IRON)
+		do_sparks(number = 2, cardinal_only = FALSE, source = src)
+		//The number of wires a bullet will randomly cut
+		var/number_of_wires_bullet_cuts = 3
+		for (var/wire = 1 to number_of_wires_bullet_cuts)
+			src.wires.cut_random_dont_mend()
```
Comment (Fikou) [inline code/game/machinery/doors/airlock.dm:1606]:
```suggestion
		for (var/i in 1 to number_of_wires_bullet_cuts)
			wires.cut_random_dont_mend()
```

**Comment by Yobrocharlie** [comment]
Ah yes I can’t wait for the 0.45 glockroach bullet to instantly destroy an APC causing me like ten minutes of suffering as I try and fail to rebuild the APC.

**Comment by Yobrocharlie** [comment]
Who needs to have any other useful tools like a door remote when you can just use gun for a decent chance to bolt doors?

**Comment by GuillaumePrata** [comment]
> Bullets will set air alarms to a random mode, then outright destroy them

If you destroy the air alarm every atmos device in the room will stop working. It could be changed to cutting a random wire like airlocks tho, it can trigger panic syphon.

*BUT* that would be kinda bad, sometimes you want to shoot people without having a space suit and most antags already use atmos to murderbone in a really non interactive way.
Adding even more to the pile seems unwise to me.

**Comment by carlarctg** [comment]
Ezxcellent idea ignore thehaters.

Does it accidentally hit these machines or do you need to aim carefully?

**Comment by Lufferly** [comment]
> Does it accidentally hit these machines or do you need to aim carefully?

Unfortunately for any non-dense machine you need to aim carefully, however can accidently hit airlocks.

**Comment by Lufferly** [comment]
wgaht

image

File: code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm [hunk 3]
```diff
@@ -690,4 +690,23 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/airalarm, 27)
...
+	var/list/possible_random_modes = list(
+		/datum/air_alarm_mode/filtering,
+		/datum/air_alarm_mode/contaminated,
+		/datum/air_alarm_mode/draught,
+		/datum/air_alarm_mode/refill,
+		/datum/air_alarm_mode/cycle,
+		/datum/air_alarm_mode/siphon,
+		/datum/air_alarm_mode/panic_siphon,
+		/datum/air_alarm_mode/off,
+		/datum/air_alarm_mode/flood,
```
Comment (nikothedude) [inline code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:704]: flood and vent siphon are usually locked behind emagging it fyi

**Comment by Pickle-Coding** [comment]
This seems cool, though I think some of these interactions could be generalised from taking sufficient damage rather than being a special bullet reaction.

File: code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm (hunk 3 above)
Comment (Lufferly) [reply code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:704]: ag good catch thanks

**Comment by Dorsisdwarf** [comment]
APCs being **instantly** destroyed by any bullet over 10 damage hitting them seems really annoying given how slow and laborious they are to repair, and kinda unbalanced since you have to whack them like 5 times with a 25 damage weapon to achieve the same.

**Comment by CoiledLamb** [comment]
this seems rlly overtuned tbh

**Comment by zxaber** [comment]
Instantly destroying APCs (presumably with a singular bullet) seems a bit much. Cool idea but possibly a bit too aggressive about it.

**Comment by DrDiasyl** [comment]
Ballistics are already powerful enough, especially when compared to lasers.

We don't need to make them even more powerful.

**Comment by Loafin34** [comment]
id say just make the APC thing only happen on 30 dam-or-up weapons, like the makarov or the l6 saw. this way you get the cool interactions (it makes sense that shooting an airlock would cut a wire, and air alarms arent exactly guarded) and keep the dumb shit out, like a wt-440 or a shotgun shell disabling it.

**Comment by optimumtact** [comment]
feature coders

**Comment by timothymtorres** [comment]
The APC thing should just cut a wire or break the panel, not outright destroy it.

**Comment by ATH1909** [comment]
don't we already have a variable for giving projectiles a damage multiplier vs. objects (and borgs, which could be a concern)

**Comment by githubuser4141** [comment]
> don't we already have a variable for giving projectiles a damage multiplier vs. objects (and borgs, which could be a concern, actually)

Yes but it also increases damage to vehicles which already have their armor against ballistics and shouldn't be affected by something like this because they literally have designated armor already and this PR is clearly targeting machinery that is not designed to be bullet resistant ugh.

**Comment by moocowswag** [comment]
> Ah yes I can’t wait for the 0.45 glockroach bullet to instantly destroy an APC causing me like ten minutes of suffering as I try and fail to rebuild the APC.

I literally made apcs easy to rebuild though why does nobody examine apcs >,<

**Comment by moocowswag** [comment]
> * Bullets will cut random wires if they hit an airlock
Plasteel on wires will protect them right?

**Comment by Lufferly** [comment]
> Plasteel on wires will protect them right?

Yes, iron will not however

**Comment by moocowswag** [comment]
PR seems fine to me, altho theoretically you can insta crit anyone standing next to an apc if you hotwire the sm....thats going to be a mixture of a mapping / skill issue until people learn about that interaction.

**Comment by Lufferly** [comment]
> PR seems fine to me, altho theoretically you can insta crit anyone standing next to an apc if you hotwire the sm....thats going to be a mixture of a mapping / skill issue until people learn about that interaction.

I wanted to avoid that specific problem so I made the damage from the apc burst constant, from my testing with my numbers it often stuns those in the radius ontop of ~20 burn, and if the electricity jumps further it does a little more than 5  burn without a stun.

**Comment by moocowswag** [comment]
lol a stun, I feel like thats alittle inconsistent since most shocks from the powernet are not going to stun unless you make physical contact with the shocked item (grilles/airlocks) with ranged ones (Tesla coils, and tesla arcs from grilles being hit) dont stun.

**Comment by LemonInTheDark** [comment]
I'm actually pretty on board with this, though I think the apc/air alarm and fire alarm interactions are too much.
They vibe differently, seem to imply an emp like effect. Don't like it, and it seems like it'd cause too much rando stuff.
I'd like the rest though, are you still up for working on this?

File: code/__DEFINES/combat.dm [hunk 4]
```diff
@@ -294,6 +294,8 @@ GLOBAL_LIST_INIT(arm_zones, list(BODY_ZONE_L_ARM, BODY_ZONE_R_ARM))
 
 #define NICE_SHOT_RICOCHET_BONUS 10 //if the shooter has the NICE_SHOT trait and they fire a ricocheting projectile, add this to the ricochet chance and auto aim angle
 
+#define BULLET_MINIMUM_DAMAGE_NEEDED_TO_EXECUTE_MACHINERY 8 //Minimum damage physical bullets need to do to trigger special effects and/or destroy on some machines
```
Comment (LemonInTheDark) [inline code/__DEFINES/combat.dm:297]: this is not a good name. it's too long

File: code/game/machinery/computer/_computer.dm [hunk 5]
```diff
@@ -21,6 +21,7 @@
 	var/authenticated = FALSE
 
 /datum/armor/machinery_computer
+	bullet = -50
```
Comment (LemonInTheDark) [inline code/game/machinery/computer/_computer.dm:24]: does this like, work?

File: code/game/machinery/doors/airlock.dm [hunk 6]
```diff
@@ -1591,6 +1591,23 @@
...
+	if (bullet.armor_flag != BULLET)
+		return ..()
+
+	// Dont cut the wires if the airlock has been reinforced with something above iron
+	if (bullet.damage > BULLET_MINIMUM_DAMAGE_NEEDED_TO_EXECUTE_MACHINERY && security_level <= AIRLOCK_SECURITY_IRON)
+		do_sparks(number = 2, cardinal_only = FALSE, source = src)
+		//The number of wires a bullet will randomly cut
+		var/number_of_wires_bullet_cuts = 3
+		for (var/i in 1 to number_of_wires_bullet_cuts)
+			wires.cut_random_dont_mend()
```
Comment (LemonInTheDark) [inline code/game/machinery/doors/airlock.dm:1606]: why is this a var

File: code/game/machinery/doors/airlock.dm [hunk 7]
```diff
@@ -1591,6 +1591,23 @@
...
 		return "photo_[frame_state]"
 
+//If a bullet does enough damage, it randomly cuts some wires
+/obj/machinery/door/airlock/bullet_act(obj/projectile/bullet)
+	// Energy attacks dont do anything special
+	if (bullet.armor_flag != BULLET)
+		return ..()
+
+	// Dont cut the wires if the airlock has been reinforced with something above iron
+	if (bullet.damage > BULLET_MINIMUM_DAMAGE_NEEDED_TO_EXECUTE_MACHINERY && security_level <= AIRLOCK_SECURITY_IRON)
```
Comment (LemonInTheDark) [inline code/game/machinery/doors/airlock.dm:1601]: early return

File: code/game/objects/items/devices/radio/intercom.dm [hunk 8]
```diff
@@ -213,6 +213,18 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/item/radio/intercom, 26)
...
+	// Energy attacks dont do anything special
+	if (bullet.armor_flag != BULLET)
+		return ..()
+
+	//If the bullet does enough damage, knock the intercom down
+	if (bullet.damage >= BULLET_MINIMUM_DAMAGE_NEEDED_TO_EXECUTE_MACHINERY)
+		do_sparks(number = 2, cardinal_only = FALSE, source = src)
+		knock_down()
+	else
+		return ..()
```
Comment (LemonInTheDark) [inline code/game/objects/items/devices/radio/intercom.dm:226]: early return

File: code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm [hunk 9]
```diff
@@ -690,4 +690,27 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/airalarm, 27)
...
+		/datum/air_alarm_mode/off,
+	)
+
+	if(obj_flags & EMAGGED)
+		possible_random_modes += list(
+			/datum/air_alarm_mode/flood,
+			/datum/air_alarm_mode/vent_siphon,
+		)
+
+	var/datum/air_alarm_mode/chosen_mode = possible_random_modes[rand(1, possible_random_modes.len)]
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm:712]: remove this but also bro just use pick

File: code/datums/wires/_wires.dm [hunk 10]
```diff
@@ -169,6 +169,19 @@
...
 	cut(wires[rand(1, wires.len)], source)
 
+//Randomly cut a wire and but dont try to mend any
+/datum/wires/proc/cut_random_dont_mend(source)
+	//Create a list of all uncut wires
+	var/list/uncut_wires = list()
+	for (var/this_wire in wires)
+		if (this_wire in cut_wires)
+			continue
+		uncut_wires += this_wire
```
Comment (LemonInTheDark) [inline code/datums/wires/_wires.dm:179]: this file is a lot longer then it needs to be
```
var/list/uncut = wires - cut_wires
if(length(uncut))
    cut(pick(uncut), source)
```
will serve you well

File: code/game/objects/structures/window.dm [hunk 11]
```diff
@@ -470,6 +471,7 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/window/unanchored/spawner, 0)
 //2023 ONE YEAR TO GO! -LT3
 /datum/armor/window_reinforced
 	melee = 80
+	bullet = -75
```
Comment (LemonInTheDark) [inline code/game/objects/structures/window.dm:474]: wtf

File: code/modules/power/apc/apc_attack.dm [hunk 12]
```diff
@@ -153,7 +153,7 @@
 			if(do_after(user, 20, target = src)) // replacing cover is quicker than replacing whole frame
 				balloon_alert(user, "cover replaced")
 				qdel(attacking_object)
-				update_integrity(30) //needs to be welded to fully repair but can work without
+				update_integrity(max(30, atom_integrity)) //needs to be welded to fully repair but can work without
```
Comment (LemonInTheDark) [inline code/modules/power/apc/apc_attack.dm:156]: why?

File: code/modules/power/apc/apc_main.dm [hunk 13]
```diff
@@ -716,3 +715,22 @@
...
 	icon_state = "power_mod"
 	desc = "Heavy-duty switching circuits for power control."
+
+//If the bullet does enough damage, shock everyone in a 3 tile radius and break
+/obj/machinery/power/apc/bullet_act(obj/projectile/bullet)
+	// Energy attacks dont do anything special
+	if (bullet.armor_flag != BULLET)
+		return ..()
+
+	if (bullet.damage >= BULLET_MINIMUM_DAMAGE_NEEDED_TO_EXECUTE_MACHINERY)
```
Comment (LemonInTheDark) [inline code/modules/power/apc/apc_main.dm:725]: early return

File: code/modules/power/apc/apc_main.dm [hunk 14]
```diff
@@ -716,3 +715,22 @@
...
+		//break and spark
+		do_sparks(number = 7, cardinal_only = FALSE, source = src)
+		atom_break()
```
Comment (LemonInTheDark) [inline code/modules/power/apc/apc_main.dm:728]: don't break

File: code/modules/power/apc/apc_main.dm [hunk 15]
```diff
@@ -716,3 +715,22 @@
 	name = "power control module"
 	icon_state = "power_mod"
 	desc = "Heavy-duty switching circuits for power control."
+
+//If the bullet does enough damage, shock everyone in a 3 tile radius and break
```
Comment (LemonInTheDark) [inline code/modules/power/apc/apc_main.dm:719]: frankly this is too much just the sparks is fine

File: code/game/machinery/computer/_computer.dm (hunk 5 above)
Comment (Pickle-Coding) [reply code/game/machinery/computer/_computer.dm:24]: I don't think I accounted for negative armour when doing armour penetration calculation changes. I should probably make sure that armour penetration won't reduce negative armour.

**Comment by Lufferly** [comment]
yeah Im still down to work on it, though it will probably be another pr and later, thanks for the review.

**Pull request closed without merging**